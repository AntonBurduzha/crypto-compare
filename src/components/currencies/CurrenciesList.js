// @flow
import * as React from 'react';
import { Row, Col, Card } from 'antd';
import { CRYPRO_COMPARE_OLD_API } from '../../constants';
import type { Currency } from '../../types/entities';
import type { Element } from 'react';

type Props = {
  list: Array<Currency>
};

const CurrenciesList = (props: Props): Element<any> =>
  <Row type="flex" justify="start" gutter={16} style={{margin: '0', padding: '0 20px' }}>
    { props.list.map((item: Currency): Element<any> => {
        return (
          <Col span={3} key={item.Id} style={{ margin: '10px 0', padding: '0' }}>
            <Card
              hoverable
              style={{ width: 140, textAlign: 'center' }}
              cover={
                <a href={`${CRYPRO_COMPARE_OLD_API}${item.Url}`} target="_blank">
                  <img alt={item.FullName} src={`${CRYPRO_COMPARE_OLD_API}${item.ImageUrl}`}/>
                </a>
              }
            >
              <Card.Meta title={item.Symbol} description={item.CoinName}/>
            </Card>
          </Col>
        );
      })
    }
  </Row>

export default CurrenciesList;
