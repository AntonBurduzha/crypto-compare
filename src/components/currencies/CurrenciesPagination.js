import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from'redux';
import * as currenciesActions from '../../actions';
import { PAGE_SIZE } from '../../constants';
import { Pagination } from 'antd';

class CurrenciesPagination extends Component {
  onChangePage = pageNumber => this.props.getNextPage(pageNumber);

  render() {
    const { fullList, page, filteredList, searchedKey } = this.props.currencies;
    const list = searchedKey ? filteredList : fullList;
    return (
      <Fragment>
        { list.length &&
            <Pagination
              total={list.length}
              showTotal={(total, range) => {
                return `${range[0]}-${range[1]} of ${total} items`}
              }
              pageSize={PAGE_SIZE}
              defaultCurrent={1}
              current={page}
              onChange={this.onChangePage}
            />
          }
      </Fragment>
    );
  }
}

function mapStateToProps(state) {
  return { currencies: state.app.currencies };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ ...currenciesActions }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(CurrenciesPagination);
