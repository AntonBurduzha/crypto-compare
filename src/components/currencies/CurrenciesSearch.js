// @flow
import * as React from 'react';
import { Input } from 'antd';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as currenciesActions from '../../actions';
import type { Action } from '../../types/actions';

type DispatchProps = { searchCryptoCurrencies: (string) => void };

export class CurrenciesSearch extends React.Component<DispatchProps> {
  onSearch = (value: string): void => this.props.searchCryptoCurrencies(value);

  render() {
    return (
      <Input.Search
        placeholder="input search text"
        enterButton="Search"
        size="large"
        onSearch={this.onSearch}
      />
    );
  }
}

export function mapDispatchToProps(dispatch: Action => void): DispatchProps {
  return bindActionCreators({ ...currenciesActions }, dispatch);
}

export default connect(null, mapDispatchToProps)(CurrenciesSearch);
