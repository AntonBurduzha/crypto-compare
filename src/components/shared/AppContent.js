// @flow
import * as React from 'react';
import { asyncComponent } from 'react-async-component';
import { NAV_TABS } from '../../constants';

const AsyncCurrenciesView = asyncComponent({ resolve: () => import('../currencies/CurrenciesView') });
const AsyncMarketBTC = asyncComponent({ resolve: () => import('../currency-market/MarketBTC') });
const AsyncMarketETH = asyncComponent({ resolve: () => import('../currency-market/MarketETH') });

const AppContent = ({ tab }: { tab: string }) => {
  switch (tab) {
    case NAV_TABS.LIST:
      return <AsyncCurrenciesView/>
    case NAV_TABS.MARKET_BTC:
      return <AsyncMarketBTC/>
    case NAV_TABS.MARKET_ETH:
      return <AsyncMarketETH/>
    default:
  }
};

export default AppContent;
