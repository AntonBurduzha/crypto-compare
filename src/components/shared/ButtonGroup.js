import React from 'react';
import { Button } from 'antd';

export const ButtonGroup = ({ startConnection, stopConnection, isConnected }) => {
  return (
    <div className="cc-socket-wrapper">
      <Button
        type="primary"
        size="large"
        style={{ width: '120px', height: '40px', marginRight: '20px' }}
        onClick={startConnection}
        disabled={isConnected}
      >
        Start
      </Button>
      <Button
        type="danger"
        size="large"
        style={{ width: '120px', height: '40px' }}
        onClick={stopConnection}
        disabled={!isConnected}
      >
        Stop
      </Button>
    </div>
  );
}
