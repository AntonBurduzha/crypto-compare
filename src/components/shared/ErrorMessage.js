// @flow
import React from 'react';
import { Row, Col } from 'antd';
import type { Element } from 'react';

export const ErrorMessage = (): Element<any> => {
  return (
    <Row type="flex" justify="center" align="middle">
      <Col span={12}>
        <h3 style={{ textAlign: 'center' }}>Oups! Error occurs, please try again later.</h3>
      </Col>
    </Row>
  );
}
