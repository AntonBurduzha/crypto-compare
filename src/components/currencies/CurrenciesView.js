// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Spin, Row, Col } from 'antd';
import * as currenciesActions from '../../actions';
import { ErrorMessage } from '../shared/ErrorMessage';
import CurrenciesPagination from './CurrenciesPagination';
import CurrenciesList from './CurrenciesList';
import CurrenciesSearch from './CurrenciesSearch';
import type { Action } from '../../types/actions';
import type { StoreState } from '../../types/reducers';
import type { Currency } from '../../types/entities';

import { currentPageListSelector } from '../../selectors';

type StateProps = { fetching: boolean, error: boolean, pageList: Array<Currency> };

type DispatchProps = { fetchCryptoCurrencies: () => void };

type Props = StateProps & DispatchProps;

export class CurrenciesView extends React.Component<Props> {
  componentDidMount() {
    this.props.fetchCryptoCurrencies();
  }

  render() {
    const { fetching, error, pageList } = this.props;
    return (
      <Spin tip="Loading..." spinning={fetching}>
        <div className="main-container">
          <Row type="flex" justify="start" align="middle" style={{ padding: '20px 30px 0px 20px' }}>
            <Col span={16}>
              <CurrenciesPagination/>
            </Col>
            <Col span={8}>
              <CurrenciesSearch/>
            </Col>
          </Row>
          <CurrenciesList list={pageList} />
          { error && <ErrorMessage /> }
        </div>
      </Spin>
    );
  }
}

export function mapStateToProps(state: StoreState): StateProps {
  return {
    fetching: state.app.currencies.fetching,
    error: state.app.currencies.error,
    pageList: currentPageListSelector(state)
  };
}

export function mapDispatchToProps(dispatch: Action => void): DispatchProps {
  return bindActionCreators({ ...currenciesActions }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(CurrenciesView);
