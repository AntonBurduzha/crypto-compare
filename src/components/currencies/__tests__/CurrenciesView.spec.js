import React from 'react';
import { Spin } from 'antd';
import { shallow } from 'enzyme';
import { CurrenciesView, mapStateToProps, mapDispatchToProps } from '../CurrenciesView';
import { ErrorMessage } from '../../shared/ErrorMessage';
import CurrenciesPagination from '../CurrenciesPagination';
import CurrenciesList from '../CurrenciesList';
import CurrenciesSearch from '../CurrenciesSearch';

describe('CurrenciesView component', () => {
  const spy = jest.fn();
  const props = {
    fetchCryptoCurrencies: spy
  };
  const wrap = shallow(<CurrenciesView {...props}/>);

  it('renders without crashing', () => {
    expect(wrap).toBeDefined();
  });

  it('fetchCryptoCurrencies called', () => {
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('spinning flag matches fetching property', () => {
    expect(wrap.find(Spin).props().spinning).toEqual(true);
    wrap.setProps({ fetching: false });
    expect(wrap.find(Spin).props().spinning).toEqual(false);
  });

  it('contains ErrorMessage if api rejected call', () => {
    expect(wrap.containsMatchingElement(<ErrorMessage/>)).toEqual(false);
    wrap.setProps({ error: true });
    expect(wrap.containsMatchingElement(<ErrorMessage/>)).toEqual(true);
    wrap.setProps({ error: false });
  });

  it('contains CurrenciesPagination', () => {
    expect(wrap.containsMatchingElement(<CurrenciesPagination/>)).toEqual(true);
  })

  it('contains CurrenciesSearch', () => {
    expect(wrap.containsMatchingElement(<CurrenciesSearch/>)).toEqual(true);
  })

  it('contains CurrenciesList', () => {
    expect(wrap.containsMatchingElement(<CurrenciesList/>)).toEqual(true);
  })

  it('has redux context', () => {
    const initState = {
      app: { currencies: { fetching: false, error: false, fullList: [1, 2, 3] } }
    };
    const dispatch = jest.fn();

    expect(mapStateToProps(initState).fetching).toEqual(false);
    expect(mapStateToProps(initState).error).toEqual(false);
    expect(mapStateToProps(initState).pageList).toEqual([]);
    expect(mapDispatchToProps(dispatch).fetchCryptoCurrencies).toBeDefined();
  });

  it('matches snapshot', () => {
    expect(wrap).toMatchSnapshot();
  });
});
