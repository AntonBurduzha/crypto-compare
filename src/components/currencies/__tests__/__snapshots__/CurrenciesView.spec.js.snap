// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrenciesView component matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CurrenciesView
    error={false}
    fetchCryptoCurrencies={[Function]}
    fetching={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <div
        className="main-container"
>
        <Row
                align="middle"
                gutter={0}
                justify="start"
                style={
                        Object {
                                "padding": "20px 30px 0px 20px",
                              }
                }
                type="flex"
        >
                <Col
                        span={16}
                >
                        <Connect(CurrenciesPagination) />
                </Col>
                <Col
                        span={8}
                >
                        <Connect(CurrenciesSearch) />
                </Col>
        </Row>
        <CurrenciesList
                list={undefined}
        />
</div>,
      "prefixCls": "ant-spin",
      "size": "default",
      "spinning": false,
      "tip": "Loading...",
      "wrapperClassName": "",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Row
            align="middle"
            gutter={0}
            justify="start"
            style={
                        Object {
                                    "padding": "20px 30px 0px 20px",
                                  }
            }
            type="flex"
>
            <Col
                        span={16}
            >
                        <Connect(CurrenciesPagination) />
            </Col>
            <Col
                        span={8}
            >
                        <Connect(CurrenciesSearch) />
            </Col>
</Row>,
          <CurrenciesList
            list={undefined}
/>,
          false,
        ],
        "className": "main-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "align": "middle",
            "children": Array [
              <Col
                span={16}
>
                <Connect(CurrenciesPagination) />
</Col>,
              <Col
                span={8}
>
                <Connect(CurrenciesSearch) />
</Col>,
            ],
            "gutter": 0,
            "justify": "start",
            "style": Object {
              "padding": "20px 30px 0px 20px",
            },
            "type": "flex",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Connect(CurrenciesPagination) />,
                "span": 16,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Connect(CurrenciesSearch) />,
                "span": 8,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "list": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div
          className="main-container"
>
          <Row
                    align="middle"
                    gutter={0}
                    justify="start"
                    style={
                              Object {
                                        "padding": "20px 30px 0px 20px",
                                      }
                    }
                    type="flex"
          >
                    <Col
                              span={16}
                    >
                              <Connect(CurrenciesPagination) />
                    </Col>
                    <Col
                              span={8}
                    >
                              <Connect(CurrenciesSearch) />
                    </Col>
          </Row>
          <CurrenciesList
                    list={undefined}
          />
</div>,
        "prefixCls": "ant-spin",
        "size": "default",
        "spinning": false,
        "tip": "Loading...",
        "wrapperClassName": "",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Row
              align="middle"
              gutter={0}
              justify="start"
              style={
                            Object {
                                          "padding": "20px 30px 0px 20px",
                                        }
              }
              type="flex"
>
              <Col
                            span={16}
              >
                            <Connect(CurrenciesPagination) />
              </Col>
              <Col
                            span={8}
              >
                            <Connect(CurrenciesSearch) />
              </Col>
</Row>,
            <CurrenciesList
              list={undefined}
/>,
            false,
          ],
          "className": "main-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "align": "middle",
              "children": Array [
                <Col
                  span={16}
>
                  <Connect(CurrenciesPagination) />
</Col>,
                <Col
                  span={8}
>
                  <Connect(CurrenciesSearch) />
</Col>,
              ],
              "gutter": 0,
              "justify": "start",
              "style": Object {
                "padding": "20px 30px 0px 20px",
              },
              "type": "flex",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <Connect(CurrenciesPagination) />,
                  "span": 16,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <Connect(CurrenciesSearch) />,
                  "span": 8,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "list": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
